src/host/_stream.cpp

```diff
@@ -172,94 +180,212 @@ void WriteCharsLegacy(SCREEN_INFORMATION& screenInfo, const std::wstring_view& t
+                if (result != 1)
+                {
+                    wch = 0;
+                }
+                if (wch)
+                {
+                    lastCharWrapped = _writeCharsLegacyUnprocessed(screenInfo, { &wch, 1 }, psScrollY);
+                }
                 break;
             }
+            }
 
-            // As a special favor to incompetent apps that attempt to display control chars,
-            // convert to corresponding OEM Glyph Chars
-            const auto cp = ServiceLocator::LocateGlobals().getConsoleInformation().OutputCP;
-            const auto ch = gsl::narrow_cast<char>(*it);
-            wchar_t wch = 0;
-            const auto result = MultiByteToWideChar(cp, MB_USEGLYPHCHARS, &ch, 1, &wch, 1);
-            if (result == 1)
+            if (writer)
             {
-                _writeCharsLegacyUnprocessed(screenInfo, { &wch, 1 }, psScrollY);
+                if (wch)
+                {
+                    writer.WriteUCS2(wch);
+                }
+                if (lastCharWrapped)
+                {
+                    writer.WriteUTF8("\r\n");
+                }
             }
-        }
+
+            ++it;
+        } while (it != end && controlCharPredicate(*it));
+    }
+
+    if (writer)
+    {
+        writer.Submit();
     }
 }
 
 // This is the main entrypoint for conhost to write VT to the buffer.
 // This wrapper around StateMachine exists so that we can add the necessary ConPTY transformations.
 void WriteCharsVT(SCREEN_INFORMATION& screenInfo, const std::wstring_view& str)
 {
-    screenInfo.GetStateMachine().ProcessString(str);
+    auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
+    auto& stateMachine = screenInfo.GetStateMachine();
+    // When switch between the main and alt-buffer SCREEN_INFORMATION::GetActiveBuffer()
+    // may change, so get the VtIo reference now, just in case.
+    auto writer = gci.GetVtWriterForBuffer(&screenInfo);
+
+    stateMachine.ProcessString(str);
+
+    if (writer)
+    {
+        const auto& injections = stateMachine.GetInjections();
+        size_t offset = 0;
+
+        const auto write = [&](size_t beg, size_t end) {
+            const auto chunk = til::safe_slice_abs(str, beg, end);
+            if (WI_IsFlagSet(screenInfo.OutputMode, DISABLE_NEWLINE_AUTO_RETURN))
+            {
+                writer.WriteUTF16(chunk);
+            }
+            else
+            {
+                writer.WriteUTF16TranslateCRLF(chunk);
+            }
+        };
+
+        for (const auto& injection : injections)
+        {
+            write(offset, injection.offset);
+            offset = injection.offset;
+
+            static constexpr std::array<std::string_view, 2> mapping{ {
+                { "\x1b[?1004h\x1b[?9001h" }, // RIS: Focus Event Mode + Win32 Input Mode
+                { "\033[?1004h" } // DECSET_FOCUS: Focus Event Mode

```

Comment:
qq: why `\033` here and `\x1b` above?

Reply:
Just a copy/paste thing, since we use one over the other depending on the author/place.